src/relay/pass/fold_scale_axis.cc

```diff
@@ -0,0 +1,559 @@
+    };
+    flist_.push_back(flazy);
+  }
+};
+
+class FScaleAxisForwardTransform : private ExprMutator {
+ public:
+  // Transform expression.
+  Expr Transform(Expr expr) {
+    expected_scale_axes_ =
+        FScaleAxisForwardPrep().Prepare(expr);
+    return this->Mutate(expr);
+  }
+
+ private:
+  // Valid axes on each node.
+  std::unordered_map<const Node*, AxesSet> expected_scale_axes_;
+  std::unordered_map<const Node*, STuple> scale_memo_;
+  // If user simply call mutate,
+  // then only Expr is returned and we cannot
+  // accept outstanding scales.
+  Expr VisitExpr(const Expr& expr) final {
+    Expr res = ExprMutator::VisitExpr(expr);
+    CHECK(!scale_memo_.count(expr.get()))
+        << "Outstanding scale";
+    return res;
+  }
+
+  STuple GetSTuple(const Expr& expr) {
+    Expr res = ExprMutator::VisitExpr(expr);
+    auto it = scale_memo_.find(expr.get());
+    if (it != scale_memo_.end()) {
+      CHECK(it->second->value.same_as(res));
+      return it->second;
+    } else {
+      auto node = make_node<STupleNode>();
+      node->value = res;
+      return STuple(node);
+    }
+  }
+
+  Expr VisitExpr_(const CallNode* call_node) final {
+    static const auto& ftransform =
+        Op::GetAttr<FForwardTransform>("FScaleAxisForwardTransform");
+    auto new_op = this->Mutate(call_node->op);
+    bool has_scale = false;
+    bool unchanged = call_node->op.same_as(new_op);
+
+    Array<STuple> call_sargs;
+    Array<Expr> call_args;
+    for (auto arg : call_node->args) {
+      STuple new_sarg = this->GetSTuple(arg);
+      unchanged &= new_sarg->value.same_as(arg);
+      if (new_sarg->axes.defined()) has_scale = true;
+      call_sargs.push_back(new_sarg);
+      call_args.push_back(new_sarg->value);
+    }
+
+    // get expected scale axes.
+    AxesSet expected_out_axes;
+    auto axis_it = expected_scale_axes_.find(call_node);
+    if (axis_it != expected_scale_axes_.end()) {
+      expected_out_axes = axis_it->second;
+    }
+    // propagation function
+    auto f = GetFunc(ftransform, call_node->op);
+    if (f != nullptr) {
+      STuple sret = f(GetRef<Call>(call_node), expected_out_axes, call_sargs);
+      if (sret.defined()) {
+        if (sret->axes.defined()) {
+          scale_memo_[call_node] = sret;
+        }
+        return sret->value;
+      }
+    }
+    // normal path
+    CHECK(!has_scale) << "Outstanding scale, on op=" << call_node->op;
+    if (unchanged) {
+      return GetRef<Expr>(call_node);
+    } else {
+      return CallNode::make(
+          new_op, call_args, call_node->attrs, call_node->type_args);
+    }
+  }
+};
+
+//----------------------------------------------
+// Per operator defs for FScaleAxisForward
+//----------------------------------------------
+
+// Intermediate operators
+Array<AxesSet> ReluForwardPrep(const Call& call, AxesSet out) {
+  return {out};
+}
+
+STuple ReluForwardTransform(const Call& ref_call,
+                              const AxesSet& expected_axes,
```

Comment:
Is there any special reason to pass `expected_axes`? It looks it is unused.

Reply:
expected_axes is necessary for Multiply, which is used to decide whether it should produce STuple or not(if expected_axes is None, it means there are not conv2d down the path).